File: engines/python/src/g2048/board.py
"""SPEC §1–§3, §5 — board encoding, line merge, moves and spawning.

A board is a list of 16 exponents, row-major (cell ``i = row * 4 + col``).
"""

from __future__ import annotations

UP, DOWN, LEFT, RIGHT = 0, 1, 2, 3
DIRECTIONS = (UP, DOWN, LEFT, RIGHT)
DIRECTION_LETTERS = "UDLR"
DIRECTION_NAMES = ("up", "down", "left", "right")

#: LINES[dir][line] = the four cell indices, starting at the edge tiles slide toward.
LINES = (
    tuple((c, 4 + c, 8 + c, 12 + c) for c in range(4)),
    tuple((12 + c, 8 + c, 4 + c, c) for c in range(4)),
    tuple((4 * r, 4 * r + 1, 4 * r + 2, 4 * r + 3) for r in range(4)),
    tuple((4 * r + 3, 4 * r + 2, 4 * r + 1, 4 * r) for r in range(4)),
)

HEX = "0123456789abcdefghijklmnopqrstuvwxyz"
_HEX_INDEX = {ch: i for i, ch in enumerate(HEX)}


def empty_board() -> list[int]:
    return [0] * 16


def board_to_hex(board) -> str:
    return "".join([HEX[e] for e in board])


def board_from_hex(s: str) -> list[int]:
    if len(s) != 16:
        raise ValueError(f"boardHex must be 16 chars, got {len(s)}")
    out = []
    for ch in s:
        v = _HEX_INDEX.get(ch.lower())
        if v is None:
            raise ValueError(f"invalid boardHex character {ch!r}")
        out.append(v)
    return out


def board_to_matrix(board) -> list[list[int]]:
    """Tile values (0 = empty) as ``m[row][col]``."""
    return [[0 if board[r * 4 + c] == 0 else 1 << board[r * 4 + c] for c in range(4)] for r in range(4)]


def board_from_matrix(m) -> list[int]:
    if len(m) != 4 or any(len(row) != 4 for row in m):
        raise ValueError("board matrix must be 4x4")
    out = []
    for row in m:
        for v in row:
            v = int(v)
            if v == 0:
                out.append(0)
            elif v >= 2 and v & (v - 1) == 0:
                out.append(v.bit_length() - 1)
            else:
                raise ValueError(f"invalid tile value {v}")
    return out


def max_exponent(board) -> int:
    return max(board)


def max_tile(board) -> int:
    e = max(board)
    return 0 if e == 0 else 1 << e


def empty_count(board) -> int:
    return board.count(0)


# --------------------------------------------------------------------------- line merge
# Memoised merge of one line keyed by the packed exponents (6 bits per cell).
# Value: (o0, o1, o2, o3, gained, changed)

_MERGE: dict = {}


def merge_line(a: int, b: int, c: int, d: int) -> tuple:
    """SPEC §3 merge of one line ordered toward the slide edge."""
    key = a | (b << 6) | (c << 12) | (d << 18)
    r = _MERGE.get(key)
    if r is None:
        tiles = [x for x in (a, b, c, d) if x]
        out = []
        gained = 0
        i = 0
        n = len(tiles)
        while i < n:
            if i + 1 < n and tiles[i] == tiles[i + 1]:
                e = tiles[i] + 1
                out.append(e)
                gained += 1 << e
                i += 2
            else:
                out.append(tiles[i])
                i += 1
        out += [0] * (4 - len(out))
        r = (out[0], out[1], out[2], out[3], gained, (out[0], out[1], out[2], out[3]) != (a, b, c, d))
        _MERGE[key] = r
    return r


def move_in_place(board: list, d: int) -> int:
    """Apply direction ``d`` in place. Returns the score gained, or -1 if invalid (board untouched)."""
    gained = 0
    changed = False
    get = _MERGE.get
    for i0, i1, i2, i3 in LINES[d]:
        a = board[i0]
        b = board[i1]
        c = board[i2]
        e = board[i3]
        r = get(a | (b << 6) | (c << 12) | (e << 18))
        if r is None:
            r = merge_line(a, b, c, e)
        if r[5]:
            board[i0] = r[0]
            board[i1] = r[1]
            board[i2] = r[2]
            board[i3] = r[3]
            gained += r[4]
            changed = True
    return gained if changed else -1


def move(board, d: int) -> tuple:
    """Return ``(board', gained, changed)``. An invalid move returns the input board unchanged."""
    b = list(board)
    g = move_in_place(b, d)
    if g < 0:
        return board, 0, False
    return b, g, True


def can_move(board, d: int) -> bool:
    for i0, i1, i2, i3 in LINES[d]:
        a = board[i0]
        b = board[i1]
        c = board[i2]
        e = board[i3]
        if b and (a == 0 or a == b):
            return True
        if c and (b == 0 or b == c):
            return True
        if e and (c == 0 or c == e):
            return True
    return False


def valid_moves(board) -> list[int]:
    return [d for d in DIRECTIONS if can_move(board, d)]


def is_over(board) -> bool:
    return not (can_move(board, 0) or can_move(board, 1) or can_move(board, 2) or can_move(board, 3))


def spawn(board: list, rng):
    """SPEC §5. Returns ``(index, exponent)`` or ``None`` if the board is full."""
    empties = [i for i, e in enumerate(board) if not e]
    if not empties:
        return None
    index = empties[rng.below(len(empties))]
    exponent = 2 if rng.below(10) == 0 else 1
    board[index] = exponent
    return (index, exponent)


def parse_direction(value) -> int:
    """Accept 0-3, 'up'/'down'/'left'/'right' or 'U'/'D'/'L'/'R'."""
    if isinstance(value, int) and not isinstance(value, bool):
        if 0 <= value <= 3:
            return value
        raise ValueError(f"invalid direction {value}")
    s = str(value).strip().lower()
    if s in DIRECTION_NAMES:
        return DIRECTION_NAMES.index(s)
    if len(s) == 1 and s.upper() in DIRECTION_LETTERS:
        return DIRECTION_LETTERS.index(s.upper())
    raise ValueError(f"invalid direction {value!r}")

File: engines/lua/g2048/board.lua
-- SPEC §1-3, §5: exponent boards, directions, moves and spawning.
--
-- A board is a Lua array of 16 exponents; cell index i (row-major, SPEC §1)
-- is stored at b[i + 1].

local M = {}

M.UP, M.DOWN, M.LEFT, M.RIGHT = 0, 1, 2, 3
M.LETTERS = "UDLR"
M.NAMES = { [0] = "up", "down", "left", "right" }

-- LINES[dir][line] = 1-based positions, starting at the edge tiles slide toward.
local LINES = {}
for dir = 0, 3 do LINES[dir] = {} end
for k = 0, 3 do
  local up, down, left, right = {}, {}, {}, {}
  for j = 0, 3 do
    up[j + 1] = j * 4 + k + 1
    down[j + 1] = (3 - j) * 4 + k + 1
    left[j + 1] = k * 4 + j + 1
    right[j + 1] = k * 4 + 3 - j + 1
  end
  LINES[0][k + 1], LINES[1][k + 1], LINES[2][k + 1], LINES[3][k + 1] = up, down, left, right
end
M.LINES = LINES

local HEX = "0123456789abcdefghijklmnopqrstuvwxyz"
local HEX_CHARS, HEX_VALUE = {}, {}
for i = 0, 35 do
  local c = HEX:sub(i + 1, i + 1)
  HEX_CHARS[i] = c
  HEX_VALUE[c] = i
  HEX_VALUE[c:upper()] = i
end

function M.empty()
  return { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 }
end

function M.copy(b)
  return { b[1], b[2], b[3], b[4], b[5], b[6], b[7], b[8], b[9], b[10], b[11], b[12], b[13], b[14], b[15], b[16] }
end

function M.equal(a, b)
  for i = 1, 16 do
    if a[i] ~= b[i] then return false end
  end
  return true
end

--- Canonical boardHex encoding.
function M.to_hex(b)
  local s = {}
  for i = 1, 16 do s[i] = HEX_CHARS[b[i]] end
  return table.concat(s)
end

--- Decode boardHex. Returns board or nil, err.
function M.from_hex(hex)
  if type(hex) ~= "string" or #hex ~= 16 then
    return nil, string.format("boardHex must be 16 chars, got %d", type(hex) == "string" and #hex or 0)
  end
  local b = {}
  for i = 1, 16 do
    local c = hex:sub(i, i)
    local v = HEX_VALUE[c]
    if not v then return nil, string.format("invalid boardHex character '%s'", c) end
    b[i] = v
  end
  return b
end

--- Decode boardHex, raising on error.
function M.must(hex)
  local b, err = M.from_hex(hex)
  if not b then error(err, 2) end
  return b
end

--- Convert a value matrix (0 = empty, powers of two) to a board.
function M.from_matrix(m)
  if type(m) ~= "table" or #m ~= 4 then return nil, "board must have 4 rows" end
  local b = M.empty()
  for r = 1, 4 do
    local row = m[r]
    if type(row) ~= "table" or #row ~= 4 then
      return nil, string.format("board row %d must have 4 cells", r - 1)
    end
    for c = 1, 4 do
      local v = math.tointeger(row[c])
      if v == nil then return nil, "invalid tile value " .. tostring(row[c]) end
      if v ~= 0 then
        if v < 2 or v & (v - 1) ~= 0 then return nil, string.format("invalid tile value %d", v) end
        local e = 0
        while v > 1 do
          v = v >> 1
          e = e + 1
        end
        if e > 35 then return nil, "tile too large" end
        b[(r - 1) * 4 + c] = e
      end
    end
  end
  return b
end

--- Tile values as matrix[row][col] (1-based).
function M.to_matrix(b)
  local m = {}
  for r = 0, 3 do
    local row = {}
    for c = 0, 3 do
      local e = b[r * 4 + c + 1]
      row[c + 1] = e ~= 0 and (1 << e) or 0
    end
    m[r + 1] = row
  end
  return m
end

function M.max_exponent(b)
  local m = 0
  for i = 1, 16 do
    if b[i] > m then m = b[i] end
  end
  return m
end

--- Largest tile value (0 for an empty board).
function M.max_tile(b)
  local e = M.max_exponent(b)
  if e == 0 then return 0 end
  return 1 << e
end

--- Apply dir in place. Returns the score gained, or -1 if the move is
--- invalid (board untouched).
function M.move_in_place(b, dir)
  local gained = 0
  local changed = false
  local lines = LINES[dir]
  for l = 1, 4 do
    local idx = lines[l]
    local p1, p2, p3, p4 = idx[1], idx[2], idx[3], idx[4]
    local a1, a2, a3, a4 = b[p1], b[p2], b[p3], b[p4]
    -- compact non-zero tiles
    local t1, t2, t3, t4 = 0, 0, 0, 0
    local n = 0
    if a1 ~= 0 then n = 1; t1 = a1 end
    if a2 ~= 0 then n = n + 1; if n == 1 then t1 = a2 else t2 = a2 end end
    if a3 ~= 0 then
      n = n + 1
      if n == 1 then t1 = a3 elseif n == 2 then t2 = a3 else t3 = a3 end
    end
    if a4 ~= 0 then
      n = n + 1
      if n == 1 then t1 = a4 elseif n == 2 then t2 = a4 elseif n == 3 then t3 = a4 else t4 = a4 end
    end
    if n > 0 then
      local r1, r2, r3, r4 = 0, 0, 0, 0
      if t1 == t2 then -- t1 ~= 0 because n > 0
        r1 = t1 + 1
        gained = gained + (1 << r1)
        if t3 ~= 0 and t3 == t4 then
          r2 = t3 + 1
          gained = gained + (1 << r2)
        else
          r2, r3 = t3, t4
        end
      else
        r1 = t1
        if t2 ~= 0 and t2 == t3 then
          r2 = t2 + 1
          gained = gained + (1 << r2)
          r3 = t4
        else
          r2 = t2
          if t3 ~= 0 and t3 == t4 then
            r3 = t3 + 1
            gained = gained + (1 << r3)
          else
            r3, r4 = t3, t4
          end
        end
      end
      if r1 ~= a1 or r2 ~= a2 or r3 ~= a3 or r4 ~= a4 then
        b[p1], b[p2], b[p3], b[p4] = r1, r2, r3, r4
        changed = true
      end
    end
  end
  if not changed then return -1 end
  return gained
end

--- Returns (new board, gained, changed) without modifying b.
function M.move(b, dir)
  local nb = M.copy(b)
  local g = M.move_in_place(nb, dir)
  if g < 0 then return M.copy(b), 0, false end
  return nb, g, true
end

--- Whether dir changes the board.
function M.can_move(b, dir)
  local lines = LINES[dir]
  for l = 1, 4 do
    local idx = lines[l]
    local prev = b[idx[1]]
    for k = 2, 4 do
      local cur = b[idx[k]]
      if cur ~= 0 and (prev == 0 or prev == cur) then return true end
      prev = cur
    end
  end
  return false
end

--- Valid directions in canonical order.
function M.valid_moves(b)
  local out = {}
  for d = 0, 3 do
    if M.can_move(b, d) then out[#out + 1] = d end
  end
  return out
end

function M.is_over(b)
  for d = 0, 3 do
    if M.can_move(b, d) then return false end
  end
  return true
end

--- Spawn a tile (SPEC §5). Returns (index, exponent) with a 0-based index,
--- or nil if the board was full (no draws).
function M.spawn(b, rng)
  local n = 0
  for i = 1, 16 do
    if b[i] == 0 then n = n + 1 end
  end
  if n == 0 then return nil end
  local k = rng:below(n) -- position among the empties, ascending
  local pos = 0
  for i = 1, 16 do
    if b[i] == 0 then
      if k == 0 then
        pos = i
        break
      end
      k = k - 1
    end
  end
  local e = 1
  if rng:below(10) == 0 then e = 2 end
  b[pos] = e
  return pos - 1, e
end

--- Parse a direction: names ("left"), letters ("L") or digits ("2").
function M.parse_direction(s)
  local t = tostring(s):lower():match("^%s*(.-)%s*$")
  for i = 0, 3 do
    if t == M.NAMES[i] then return i end
  end
  if #t == 1 then
    local i = ("udlr"):find(t, 1, true)
    if i then return i - 1 end
    if t >= "0" and t <= "3" then return tonumber(t) end
  end
  return nil, string.format("invalid direction '%s'", tostring(s))
end

return M
